File: Cpp/5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
/*
manacher's algo: https://github.com/Yukinichi/leetcode/blob/master/Note/Manacher_Algorithm.md
time:O(n)
space: O(n)
*/
public:
    string longestPalindrome(string s) {
        if (s.size() < 2) return s;
        string p = "^";
        for (auto c : s) {
            p += "#";
            p += c;
        }
        p += "#$";
        vector<int> dp(p.size(), 0);
        int center = 0, r = 0;  // r: center对应最长回文右边界
        for (int i = 1; i < p.size() - 1; ++i) {
            int i_m = 2 * center - i;
            dp[i] = r >= i ? min(r - i, dp[i_m]) : 0;
            while (p[i + dp[i] + 1] == p[i - dp[i] - 1]) { dp[i]++; }
            if (i + dp[i] > r) {
                center = i;
                r = i + dp[i];
            }
        }
        center = max_element(dp.begin(), dp.end()) - dp.begin();
        return s.substr((center - dp[center]) / 2, dp[center]);
    }
};
// @lc code=end

File: src/5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    /*
    start from the middle and expand the palindrom string
    time: O(n^2)
    space: O(n)
    */
    int max = 0;
    String res = "";
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0) return res;
        for (int i = 0; i < s.length(); ++i) {
            extend(s,i,i);
            extend(s,i,i + 1);
        }
        return res;
    }

    private void extend(String s, int left, int right) {
        while (left >=0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        if (right - left + 1 > max) {
            max = right - left + 1;  // we need to track r-l+1 instead of r-l-1, for "abcd", l = 1,r=2,max=0
            res = s.substring(left + 1,right);  // current left != right
        }
    }
}
// @lc code=end

